src/expr/src/interpret.rs

```diff
@@ -898,41 +898,74 @@ impl<'a> Interpreter for ColumnSpecs<'a> {
-/// or it's been special-cased in the interpreter.) And we say a function-call expression is
-/// pushdownable overall if at least one of its arguments is pushdownable, and the function is
-/// is pushdownable in that argument. For example, `3 + #0` is pushdownable (because the second
-/// argument to `+` is pushdownable and `+` is monotone) but `3 / #0` is not (because while the
-/// second argument is pushdownable, `/` is not monotone on the right-hand side).
+/// In practice, we internally need to distinguish between expressions that are trivially predicable
+/// (because they're constant) and expressions that depend on the column ranges themselves.
+/// See the [TraceSummary] variants for those distinctions, and [TraceSummary::pushdownable] for
+/// the overall assessment.
 #[derive(Debug)]
 pub struct Trace;
 
+/// A summary type for the [Trace] interpreter.
+// NB: the ordering here is important!
+#[derive(Copy, Clone, Debug, PartialOrd, PartialEq, Ord, Eq)]
+pub enum TraceSummary {
+    /// The expression is constant: we can evaluate it without any runtime information.
+    Constant,
+    /// The expression depends on runtime information, but in "predictable" way... ie. if we know
+    /// the range of possible values for all columns and unmaterializable functions, we can
+    /// predict the possible values of the output.
+    Dynamic,
+    /// The expression depends on runtime information in an unpredictable way.
+    Unknown,
+}
+
+impl TraceSummary {
+    /// We say that a function is "pushdownable" for a particular
+    /// argument if `ColumnSpecs` can determine the spec of the function's output given the input spec for
+    /// that argument. (In practice, this is true when either the function is monotone in that argument
+    /// or it's been special-cased in the interpreter.)
+    fn apply_fn(self, pushdownable: bool) -> Self {
+        match self {
+            TraceSummary::Constant => TraceSummary::Constant,
+            TraceSummary::Dynamic => match pushdownable {
+                true => TraceSummary::Dynamic,
+                false => TraceSummary::Unknown,
+            },
+            TraceSummary::Unknown => TraceSummary::Unknown,
+        }
+    }
+
+    /// We say that an expression is "pushdownable" if it's either constant or dynamic.
+    pub(crate) fn pushdownable(self) -> bool {
+        match self {
+            TraceSummary::Constant | TraceSummary::Dynamic => true,
+            TraceSummary::Unknown => false,
+        }
+    }
+}
+
 impl Interpreter for Trace {
-    type Summary = bool;
+    type Summary = TraceSummary;
 
     fn column(&self, _id: usize) -> Self::Summary {
-        true
+        TraceSummary::Dynamic
     }
 
     fn literal(&self, _result: &Result<Row, EvalError>, _col_type: &ColumnType) -> Self::Summary {
-        false
+        TraceSummary::Constant
     }
 
     fn unmaterializable(&self, _func: &UnmaterializableFunc) -> Self::Summary {
-        false
+        TraceSummary::Dynamic
```

Comment:
Why is this changing from `false` to `Dynamic` and not to `Unknown`? In which cases can you interpret an unmaterializable function to a specific value range?

Reply:
We do this for `mz_now()`... in the persist source we can bound the range based on the as-of of the query. (See [here](https://github.com/MaterializeInc/materialize/blob/f8c5f3a2939f7e3bdc3aa0909826b5fd50744ae8/src/expr/src/interpret.rs#L682-L690) for how that looks on the mz-expr side.) We could in theory fill in constant values for some of the other functions, but in practice they get compiled away by the time the expression reaches us...